packages/apps/dashboard/ui-2024/src/features/Leaderboard/hooks/useDataGrid.tsx

```diff
@@ -0,0 +1,211 @@
+import { Box, Typography } from '@mui/material';
+import { GridColDef, GridRenderCellParams } from '@mui/x-data-grid';
+import { LeaderBoardData } from '@services/api/use-leaderboard-details';
+import { useBreakPoints } from '@utils/hooks/use-is-mobile';
+import { useLeaderboardSearch } from '@utils/hooks/use-leaderboard-search';
+import HelpOutlineIcon from '@mui/icons-material/HelpOutline';
+import CustomTooltip from '@components/CustomTooltip';
+import { colorPalette } from '@assets/styles/color-palette';
+import { useMemo } from 'react';
+
+import { RoleCell } from '../components/RoleCell';
+import { AddressCell } from '../components/AddressCell';
+import { ChainCell } from '../components/ChainCell';
+import { SelectNetwork } from '../components/SelectNetwork';
+import { ReputationLabel } from '../components/ReputationLabel';
+import { TextCell } from '../components/TextCell';
+
+export const useDataGrid = (data: LeaderBoardData) => {
+	const {
+		filterParams: { chainId },
+	} = useLeaderboardSearch();
+
+	const {
+		mobile: { isMobile },
+	} = useBreakPoints();
+	const formattedData = useMemo(() => {
+		return data.map((row, index) => {
+			return {
+				...row,
+				id: index,
+			};
+		});
+	}, [data]);
+
+	const visibleRows = useMemo(() => {
+		if (chainId !== -1) {
+			return formattedData.filter((elem) => elem.chainId === chainId);
+		}
+
+		return formattedData;
+	}, [formattedData, chainId]);
+
+	const columns: GridColDef<(typeof visibleRows)[number]>[] = useMemo(
+		() => [
+			...(isMobile
+				? []
+				: [
+						{
+							field: 'id',
+							headerName: '',
+							width: 30,
+							headerClassName: 'home-page-table-header',
+							sortable: false,
+							renderCell: (params: GridRenderCellParams) => (
+								<Typography
+									variant="body1"
+									height="100%"
+									display="flex"
+									alignItems="center"
+								>
+									{params.api.state.sorting.sortedRows.indexOf(params.id) + 1}
```

Comment:
This logic is implicit and easy to break, esp. with the fact we have a hack with `onSortModelChange` for it. The right way of doing this would be to implement something similar ["server-side" sorting](https://mui.com/x/react-data-grid/sorting/#server-side-sorting), where we sort on our end and change raw index accordingly.

IMU the idea of this leaderboard is to show items depending on "amountStaked" by default and "index" column should show the "rank" based on this criteria. In case user decides to sort the table based on some other column - rank won't change.

So I see three possible ways of making it right:
1) Implement it like "server-side" to avoid hacks with re-render triggering
2) Remove this column at all
3) Display it as is, assuming it is a `rank` value (so it will work [like this table](https://www.coingecko.com/en/chains))

IMO #1 is overhead here, so I would go #3 since it corresponds to the idea of leaderboard

Reply:
Discussed with the team, let's do the following now:
- keep sorting only for `stake`, `reputation` and `fee` columns
- always keep 1-2-3 raw index, no matter what is the sort criteria; either go with a suggested option #1 or find some other ways to achieve that w/o implicit trigger of re-rendering